include/ginkgo/core/base/math.hpp

```diff
@@ -499,6 +510,108 @@ GKO_INLINE GKO_ATTRIBUTES constexpr T get_superior_power(
+    {                                                                        \
+        constexpr auto infinity = INFINITY;                                  \
+        return value == value && value != infinity && value != -infinity;    \
+    }                                                                        \
+    static_assert(true,                                                      \
+                  "This assert is used to counter the false positive extra " \
+                  "semi-colon warnings")
+#elif defined(__CUDA_ARCH__)
+#define GKO_DEFINE_ISFINITE_FOR_TYPE(_type)                                  \
+    GKO_INLINE __device__ bool isfinite(const _type &value)                  \
+    {                                                                        \
+        return ::isfinite(value);                                            \
+    }                                                                        \
+    static_assert(true,                                                      \
+                  "This assert is used to counter the false positive extra " \
+                  "semi-colon warnings")
+#else  // !defined(__CUDA_ARCH__)
+#define GKO_DEFINE_ISFINITE_FOR_TYPE(_type)                                  \
+    GKO_INLINE __host__ bool isfinite(const _type &value)                    \
+    {                                                                        \
+        using std::isfinite;                                                 \
+        return isfinite(value);                                              \
+    }                                                                        \
+    static_assert(true,                                                      \
+                  "This assert is used to counter the false positive extra " \
+                  "semi-colon warnings")
+#endif  // defined(__CUDA_ARCH__) && defined(__clang__)
+
+GKO_DEFINE_ISFINITE_FOR_TYPE(float);
+GKO_DEFINE_ISFINITE_FOR_TYPE(double);
+
+#undef GKO_DEFINE_ISFINITE_FOR_TYPE
+
+/**
+ * Checks if a component of a given complex value is positive infinity,
+ * negative infinity, or NaN
+ *
+ * @param value  complex value to check
+ *
+ * returns `true` if a component of the complex value is either positive or
+ *         negative infinity or NaN. Otherwise `false`
+ */
+#define GKO_DEFINE_ISFINITE_FOR_COMPLEX_TYPE(_type)                          \
+    GKO_INLINE GKO_ATTRIBUTES bool isfinite(const _type &value)              \
+    {                                                                        \
+        return isfinite(value.real()) || isfinite(value.imag());             \
+    }                                                                        \
+    static_assert(true,                                                      \
+                  "This assert is used to counter the false positive extra " \
+                  "semi-colon warnings")
+
+
+#if !defined(__CUDA_ARCH__)
```

Comment:
This instantiations are for complex types when not on CUDA, right ? What about for CUDA ?

Reply:
They are in [`cuda/base/math.hpp`](https://github.com/ginkgo-project/ginkgo/pull/324/files/ff7eab0396f7472c1bf70a696ed55a689ea40369#diff-845ceb62f9be5a134574d7d8619d9e81R104)

Reply:
Ah, okay. I missed that.